### PR TITLE
feat: renew session for expiration, re-execute when error

### PR DESCRIPTION
**Pull request by wey-gu**
close: #276

## What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### Description:

- renew session when E_SESSION_INVALID or E_SESSION_TIMEOUT
- re-execute for E_EXECUTION_ERROR.

### other changes

CI changed: avoid using pdm install, this was due to since py3.12, future's installation requires setuptools but setuptools was removed and thus it requires explicit installation/dependency on setuptools, while, the setuptools version cannot be resolved in py3.6.2+ actually, so we have to use pip install to mitigate this. (cc @frostming , I replaced pdm install with pip install, ping you to see if elegant way exists?)

### CODE REVIEW

File: nebula3/gclient/net/Session.py [hunk 1]
```diff
@@ -18,14 +19,22 @@
...
 
 class Session(object):
-    def __init__(self, connection, auth_result: AuthResult, pool, retry_connect=True):
+    def __init__(
+        self,
+        connection,
+        auth_result: AuthResult,
+        pool,
+        retry_connect=True,
+        retry_times=3,
```
Comment (Nicole00) [inline nebula3/gclient/net/Session.py:28]: we can add another config interval_time to sleep some time between retry executions.

Comment (wey-gu) [reply nebula3/gclient/net/Session.py:28]: Done :)

File: nebula3/gclient/net/Session.py [hunk 2]
```diff
@@ -222,6 +264,20 @@ def execute_json_with_parameter(self, stmt, params):
                     )
                     return resp_json
             raise
+        except ExecutionErrorException as eee:
+            retry_count = 0
+            while retry_count < self._retry_times:
+                try:
```
Comment (Nicole00) [inline nebula3/gclient/net/Session.py:270]: sleep before retry execution.

Comment (wey-gu) [reply nebula3/gclient/net/Session.py:270]: done!

**Comment by wey-gu** [comment]
in Python3.12, setuptools was removed by default

https://peps.python.org/pep-0632/#migration-advice

will fix the CI now.

**Review by Nicole00 — APPROVED** [review]

**Pull request merged**